lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -665,14 +688,159 @@ def write(chunk)
+                          error_code: "google.rpc.Code[#{error_code}]",
+                          error: error_message
+              end
+            end
           end
         end
       end
     end
 
+    # Extract a map of error details from an potentially partially successful
+    # request. The keys in this map are stringified indexes of the failed log
+    # entries in the original json payload, and the values are the error code
+    # and error message.
+    #
+    # A sample error.body looks like:
+    # {
+    #   "error": {
+    #     "code": 403,
+    #     "message": "User not authorized.",
+    #     "status": "PERMISSION_DENIED",
+    #     "details": [
+    #       {
+    #         "@type": "type.googleapis.com/google.logging.v2.WriteLogEntriesPar
+    #           tialErrors",
+    #         "logEntryErrors": {
+    #           "0": {
+    #             "code": 7,
+    #             "message": "User not authorized."
+    #           },
+    #           "1": {
+    #             "code": 3,
+    #             "message": "Log name contains illegal character :"
+    #           },
+    #           "3": {
+    #             "code": 3,
+    #             "message": "Log name contains illegal character :"
+    #           }
+    #         }
+    #       },
+    #       {
+    #         "@type": "type.googleapis.com/google.rpc.DebugInfo",
+    #         "detail": ...
+    #       }
+    #     ]
+    #   }
+    # }
+    #
+    # The root level "code", "message", and "status" simply match the root
+    # cause of the first failed log entry. For example, if we switched the order
+    # of the log entries, then we would get:
+    # {
+    #    "error" : {
+    #       "code" : 400,
+    #       "message" : "Log name contains illegal character :",
+    #       "status" : "INVALID_ARGUMENT",
+    #       "details": ...
+    #    }
+    # }
+    # We will ignore it anyway and look at the details instead which includes
+    # info for all failed log entries.
+    #
+    # In this example, the logEntryErrors we are extracting should be:
+    # {
+    #   "0": {
+    #     "code": 7,
+    #     "message": "User not authorized."
+    #   },
+    #   "1": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   },
+    #   "3": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   }
+    # }
+    def extract_log_entry_errors(error)
+      error_body = ensure_hash(JSON.parse(error.body))
+      error = ensure_hash(error_body['error'])
+      error_details = ensure_array(error['details'])
+      partial_errors = error_details.detect do |error_detail|
+        ensure_hash(error_detail)['@type'] == PARTIAL_ERROR_FIELD
+      end
+      unless partial_errors
+        @log.warn 'Failed to detect an error detail with type' \
+          " #{PARTIAL_ERROR_FIELD}."
+      end
+      ensure_hash(ensure_hash(partial_errors)['logEntryErrors'])
+    rescue JSON::ParserError => e
```

Comment:
We can put this `rescue` around the `construct_error_details_map(extract_log_entry_errors(error))` call instead. That way it'll also catch the errors thrown by the `ensure_hash` in `construct_error_details_map`...
We just need to remove the "`JSON parsing error`" bit from the error message, because it'll be apparent from the exception object.

Reply:
Changed.

Reply:
Sorry for being unclear -- I meant putting it around the *call*, in line 696... You can just do that in the conditional block:
```ruby
            if @partial_success
              @log.debug 'Encountered error when talking to Stackdriver' \
                         ' Logging API.',
                         error: error.class, error_body: error.body
              error_details_map = construct_error_details_map(
                extract_log_entry_errors(error))
            rescue JSON::ParserError => e
              @log.warn 'Failed to extract log entry errors from the error' \
                        " details: #{error.body}.", error: e
            end
```

Reply:
I did try this. Yet I got `Avoid more than 3 levels of block nesting` warning from `Rubocop`.

On a second thought, there is no real reason to keep the two functions separate. I just combined them so I can have one `rescue` that covers the whole function.

PTAL.

Reply:
Good point. If you only have one function, this works fine.